lemur/plugins/lemur_acme/challenge_types.py

```diff
@@ -0,0 +1,240 @@
+                    chall.append(i)
+
+        if len(chall) == 0:
+            raise Exception('HTTP-01 challenge was not offered by the CA server.')
+        else:
+            validation_target = None
+            for option in json.loads(issuer_options["authority"].options):
+                if option["name"] == "tokenDestination":
+                    validation_target = option["value"]
+
+            if validation_target is None:
+                raise Exception('No token_destination configured for this authority. Cant complete HTTP-01 challenge')
+
+        for challenge in chall:
+            response = self.deploy(challenge, acme_client, validation_target)
+
+            acme_client.answer_challenge(challenge, response)
+
+        current_app.logger.info("Uploaded HTTP-01 challenge tokens, trying to poll and finalize the order")
+
+        finalized_orderr = acme_client.finalize_order(orderr, datetime.datetime.now() + datetime.timedelta(seconds=90))
+
+        pem_certificate = OpenSSL.crypto.dump_certificate(
+            OpenSSL.crypto.FILETYPE_PEM,
+            OpenSSL.crypto.load_certificate(
+                OpenSSL.crypto.FILETYPE_PEM, finalized_orderr.fullchain_pem
+            ),
+        ).decode()
+
+        if current_app.config.get("IDENTRUST_CROSS_SIGNED_LE_ICA", False) \
+                and datetime.datetime.now() < datetime.datetime.strptime(
+                current_app.config.get("IDENTRUST_CROSS_SIGNED_LE_ICA_EXPIRATION_DATE", "17/03/21"), '%d/%m/%y'):
+            pem_certificate_chain = current_app.config.get("IDENTRUST_CROSS_SIGNED_LE_ICA")
+        else:
+            pem_certificate_chain = finalized_orderr.fullchain_pem[
+                                    len(pem_certificate):  # noqa
+                                    ].lstrip()
+
+        # validation is a random string, we use it as external id, to make it possible to implement revoke_certificate
+        return pem_certificate, pem_certificate_chain, None
+
+    def deploy(self, challenge, acme_client, validation_target):
+
+        if not isinstance(challenge.chall, challenges.HTTP01):
+            raise AcmeChallengeMissmatchError(
+                'The provided challenge is not of type HTTP01, but instead of type {}'.format(
+                    challenge.__class__.__name__))
+
+        destination = destination_service.get(validation_target)
+
+        if destination is None:
+            raise Exception(
+                'Couldn\'t find the destination with name {}. Cant complete HTTP01 challenge'.format(validation_target))
+
+        destination_plugin = plugins.get(destination.plugin_name)
+
+        response, validation = challenge.response_and_validation(acme_client.net.key)
+
+        destination_plugin.upload_acme_token(challenge.chall.path, validation, destination.options)
+        current_app.logger.info("Uploaded HTTP-01 challenge token.")
+
+        return response
+
+    def cleanup(self, challenge, acme_client, validation_target):
```

Comment:
Implementing a cleanup makes sense, I'll add that as well. But it will need a delete_acme_token function within the S3Destination to make it work with the S3 plugin

Reply:
good point, will take care of that in a separate PR.

Reply:
raised the PR https://github.com/Netflix/lemur/pull/3245